rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,71 @@
+
+To generate an event stream we'll connect to the database with a regular connection, create a temporary logical replication slot, then poll changes on it via the [`pg_logical_slot_get_changes`](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-REPLICATION) function. As the builtin `pgoutput` logical decoding plugin cannot be used without errors via the SQL interface except in very recent versions of Postgres (11.20, 12.15, 13.11, 14.8 or 15.3, released in May 2023), we use the third-party but very popular [`wal2json`](https://github.com/eulerto/wal2json) plugin, which is available by default on the managed Postgres offerings in Azure, AWS and GCP, and is installable from the Debian/Ubuntu official packages or from the Postgres APT and YUM repos.
+
+The `pg_logical_slot_get_changes` function (with the appropriate `wal2json` options) will return one entry per change, which maps nicely to the backend event model; updates and inserts will be rendered as `OpPut` events, deletes as `OpDelete` events.
+
+### Requirements and recommendations
+
+Any supported version of Postgres (11 to 15) should work, with `wal2json` 2.1 or later, although for better results, Postgres 13 or higher is recommended (as logical decoding had some internal improvements).
+
+Because the db user that Teleport connects as needs `REPLICATION` permissions, which effectively grants read access to the entire database cluster, the backend should not be used in a database cluster that also contains data that the Teleport auth server should not have access to (it's fine to use a single cluster for the Postgres backend and the Postgres audit log storage, for instance).
+
+Because of how temporary replication slots (and prepared statements) are tied to individual connections, the use of external connection poolers (such as `pgbouncer`) is not supported.
+
+For self-managed Postgres setups we recommend the use of [certificate authentication](https://www.postgresql.org/docs/current/auth-cert.html); in Azure we support and recommend the use of Azure AD authentication over the use of hardcoded passwords.
+
+Teleport will attempt to create the database specified in the configuration for itself, and set it up with the correct schema (as well as apply any future migrations, if needed). It's possible to create the database manually and not give Teleport permissions to create databases, and technically it's even possible to manually set up the database schema (to avoid granting Teleport the ability to create tables, for instance), but we recommend giving Teleport its own database for the backend - and since Teleport will need read-write access to all the data in it anyway, we also recommend letting Teleport manage the schema automatically.
+
+### UX
+
+TBD, so far it's only a `conn_string` passed straight to `pgxpool` and a boolean `azure` flag with an optional `azure_client_id` string to select a specific identity if multiple are available.
```

Comment:
Open UX questions:
- which tunables should we expose? (delete batch size, expiry batch size, expiry interval, disable expiry loop, event poll interval, event batch size)
- the conn_string (as well as the rest of the pgx/pgxpool machinery) can use a password, either embedded or by reading it from the `~/.pgpass` file. Should we document that? Should we detect it and deliberately refuse to proceed with a password? What about TLS options such as `sslmode=verify-full`?
- should we require an option for the "change feed mode" even if we currently only support `wal2json` with the SQL interface? Requiring an option to be set might be better in the long term compared to having a default that we might deprecate in the future.